### PR TITLE
Try to improve CLI parsing

### PR DESCRIPTION
**Pull request by cosmo-ray**
Try to fix this:
https://wiki.xoreos.org/index.php?title=TODO#Command_line_parsing_in_the_tools
I've try to keep backward compatibility, so the usage should be just the same as before, I was unable to test every options of every program, but the one I've try was working, I hope I've break nothing.

### CODE REVIEW

**Comment by DrMcCoy** [comment]
Neat, thanks! :)

I probably won't have time to look at it in detail before January 1st, though.

One minor thing I can see immediately: the commit message should begin with a uppercase letter after the module tag. I.e. "COMMON: Add a new CLI class" instead of "COMMON: add a new CLI class".

**Comment by cosmo-ray** [comment]
commits messages fix

**Comment by cosmo-ray** [comment]
ping

File: src/convert2da.cpp [hunk 1]
```diff
@@ -86,100 +86,35 @@ int main(int argc, char **argv) {
...
-	std::fprintf(stream, "            --version           Display version information\n");
-	std::fprintf(stream, "  -o <file> --output <file>     Write the output to this file\n");
-	std::fprintf(stream, "  -a        --2da               Convert to ASCII 2DA (default)\n");
-	std::fprintf(stream, "  -b        --2dab              Convert to binary 2DA\n");
-	std::fprintf(stream, "  -c        --csv               Convert to CSV\n\n");
-	std::fprintf(stream, "If several files are given, they must all be GDA and use the same\n");
-	std::fprintf(stream, "column layout. They will be pasted together and printed as one GDA.\n\n");
-	std::fprintf(stream, "If no output file is given, the output is written to stdout.\n");
-}
+// std::fprintf(stream, "  -o <file> --output <file>     Write the output to this file\n");
```
Comment (DrMcCoy) [inline src/convert2da.cpp]: Any reason you kept this?

File: src/fixpremiumgff.cpp [hunk 2]
```diff
@@ -76,82 +76,22 @@ bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue
...
+	using Common::Cli::NoOption;
+	using Common::Cli::makeEndArgs;
+	using Common::Cli::kContinueParsing;
+	NoOption inFileOpt(false, new ValGetter<Common::UString &>(inFile, "input files"));
+	NoOption outFileOpt(false, new ValGetter<Common::UString &>(outFile, "output files"));
+	Parser parser(argv[0], "Repair BioWare GFF files found in encrypted NWN premium module HAKs\n",
+		      "If no ID is given is given, it is guessed from the file name.",
+		      returnValue, makeEndArgs(&inFileOpt, &outFileOpt));
+
+
```
Comment (DrMcCoy) [inline src/fixpremiumgff.cpp]: Trailing whitespace

File: src/gff2xml.cpp [hunk 3]
```diff
@@ -119,145 +119,61 @@ bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue
...
+	parser.addOption("jade", "Use Jade Empire encodings", kContinueParsing,
+			 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDJade, game)));
+	parser.addOption("witcher", "Use The Witcher encodings", kContinueParsing,
+			 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDWitcher, game)));
+	parser.addOption("dragonage", "Use Dragon Age encodings", kContinueParsing,
+			 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDDragonAge, game)));
+	parser.addOption("dragonage2", "Use Dragon Age II encodings", kContinueParsing,
+			 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDDragonAge2, game)));
+	parser.addOption("encoding", "Override an encoding", kContinueParsing,
+			 new Callback<EncodingOverrides &>("str", parseEncodingOverride, encOverrides));
```
Comment (DrMcCoy) [inline src/gff2xml.cpp]: valgrind says that this line leaks, the new Callback

File: src/unerf.cpp [hunk 4]
```diff
@@ -148,149 +148,74 @@ void parsePassword(const Common::UString &arg, std::vector<byte> &password) {
...
+
+	parser.addOption("nwn2", "Alias file types according to Neverwinter Nights 2 rules",
+			 kContinueParsing,
+			 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDNWN2, game)));
+	parser.addOption("jade", "Alias file types according to Jade Empire rules",
+			 kContinueParsing,
+			 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDJade, game)));
+	parser.addOption("pass", "Decryption password, if required, in hex notation",
+			 kContinueParsing,
+			 new Callback<std::vector<byte> &>("hex", parsePassword, password));
```
Comment (DrMcCoy) [inline src/unerf.cpp]: Here too the Callback (and in the lines below) leak

**Comment by cosmo-ray** [comment]
thanks for the review and sorry for the time I take to answer.
I've try to fix all issues, I add a Parse::addSpace() method to handle groupings via newlines,
I didn't add "\n"  in any explanatory texts, but I've made Parser::process automatically prepend a "\n" to the explanatory text if the string isn't empty.

**Comment by DrMcCoy** [comment]
Very nice, thank you! :)

I merged it with this push: https://github.com/xoreos/xoreos-tools/compare/f0dc3476727c...1b6d66a9f6d4.

**Pull request closed without merging**